### ISSUE TITLE
Performance issue on `compute` with reshaped arrays

### ISSUE THREAD

**Issue by nbren12** [dask/dask#3650]
I ran into a performance issue with related to reshaped dask arrays while using xarray (https://github.com/pydata/xarray/issues/2241). I opened an issue there, but I think this is actually more of a dask issue. Here is some example code which demonstrates the slow down
```python
>>> a = da.random.random((64, 64, 64), (8,8,-1))
>>> b = a.reshape((64**2, 64))
>>> %timeit b[0].compute()
3.2 ms ± 13 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit a[0,0].compute()
628 µs ± 13.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
`b[0]` and `a[0,0]` should in principle refer to the same data, but the reshaped values take almost 4 times as long to compute. 

Edit:

This is on dask 0.17.1

**Comment by mrocklin** [comment]
I recommend looking at

    b[0].visualize('mypicture.png', optimize_graph=True)

You'll find that to produce the first chunk Dask now needs to create and
collect many other chunks, leading to an increased compute time.

**Comment by nbren12** [comment]
Ok. thanks for tips. I am surprised that dask needs to do all that work since the first chunk of `b` is just a reshaped version of the first chunk of `a`.

**Comment by shoyer** [comment]
Note the chunksizes of this arrays:
```
>>> a
dask.array<random_sample, shape=(64, 64, 64), dtype=float64, chunksize=(8, 8, 64)>
>>> b
dask.array<reshape, shape=(4096, 64), dtype=float64, chunksize=(64, 64)>
```

When you reshape the first dimensions from `(64, 64)` to `(64**2,)`, your chunks actually get scrambled: the first 64 elements of `b` (in its first chunk) read off the first row of the `a`, which includes 8 different chunks.

When computations are involved, dask doesn't have any way of telling the original `da.random.random()` that you're indexing out part of the result and it doesn't need to compute full chunks. So considerably more work is done.

**Comment by nbren12** [comment]
> your chunks actually get scrambled

Thanks for pointing that out. Now it makes sense to me. I guess the the problem is that the chunks have to be contiguous in the new shape. In principal, I guess strided chunks would fix this performance issue, but I would never expect for such a thing to be implemented in dask. Maybe this issue could be best addressed in `xarray` by using some smart indexing tricks.

@mrocklin Feel free to close this issue if you want.

**Comment by jakirkham** [comment]
Haven't given this a lot of thought. So this suggestion might not work, but have you explored doing rechunking first? If you expand the "fast axes" to the length of the shape first, things might be a bit better. Again haven't test this, but would be curious to hear what you learn from this experiment.

**Comment by tlambert03** [comment]
I know it's been a long time since this topic, but I recently ran into unexpected `reshape` behavior while using `dask.delayed` or `dask-image` to load a folder of tiffs and then trying to reshape that folder to reflect the dimensions in the underlying dataset.  For instance:

```python
from dask_image.imread import imread
stack = imread('/data/*.tif') 
stack.reshape(600, 2, 64, 256, 280)[0, 0].compute()  # triggers 2 file read events
# this is what I actually need to do to separate dims
stack.reshape(2, 600, 64, 256, 280)[0, 0].compute()  # triggers 600 file read events
```

As @mrocklin pointed out, using `.visualize()` is useful to see why...  here's a minimal example with reduced array size:

```python
from dask import delayed
import dask.array as da

@delayed
def f(a,b):
    # simulating a file-read event
    return np.zeros((10, 10))

arr = da.stack(
    [
        da.from_delayed(f(a, b), shape=(10, 10), dtype="float64")
        for a in range(10)
        for b in range(2)
    ]
)
arr.shape  # (20, 10, 10)

arr.reshape(10, 2, 10, 10)[0, 0].visualize()  # 2 calls to f()
```
![69357424-0f7eab80-0c53-11ea-8bdc-3fb6f738eb53](https://user-images.githubusercontent.com/1609449/69368651-82dee800-0c68-11ea-8370-feeb055bdf22.png)

```python
arr.reshape(2, 10, 10, 10)[0, 0].visualize()  # 10 calls to f()
```
![69349145-48178880-0c45-11ea-99ec-91776ddbd336](https://user-images.githubusercontent.com/1609449/69368659-8a05f600-0c68-11ea-8fc4-347e0b249a93.png)

I thought rechunking might help... but it doesn't:
```python
rechunked = arr.reshape(2, 10, 10, 10).rechunk((1, 1, 10, 10))
rechunked
# dask.array<rechunk-merge, shape=(2, 10, 10, 10), dtype=float64, chunksize=(1, 1, 10, 10)>

rechunked[0, 0].compute()  # still calls f() 10 times with or without optimize_graph=True
```

The only way I can think of to avoid this to avoid reshape altogether by manually constructing dask arrays for each dimension, and then stacking them.  As an example for my `dask-image` use case:
```python
from dask_image.imread import imread
stack = da.stack([imread(f"data/*ch{i}*.tif") for i in range(nchannels)])
# then calling stack[0,0].compute() only incurs one read
```
that seems a (tiny) bit harder to scale... so just wanted to make sure that's the recommended approach?  thanks!

**Comment by jni** [comment]
@shoyer I don't understand why chunks are getting "consolidated" in @tlambert03's example:

```python
In [4]: arr.chunks
Out[4]: ((1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (10,), (10,))

In [5]: arr.reshape(2, 10, 10, 10).chunks
Out[5]: ((1, 1), (10,), (10,), (10,))
```

The chunksize goes from (1, 10, 10) to (1, 10, 10, 10), so stacking of the chunks is happening automatically and, in this case, with a detrimental effect on performance. Unless I'm misunderstanding what's happening, it seems to me that reshape and other functions should gain the ability to turn off automatic rechunking, and this problem would then go away?

**Comment by shoyer** [comment]
Yeah, I don’t understand what is going on in that example. It looks like a bit of a different situation than what is described in the original issue here.

**Comment by chrisroat** [comment]
The rehsape algorithm is setup to rechunk into singleton chunks along all but the first dimension.

I created an example PR to show a shortcut when collapsing chunksize=1 reshaping:
https://github.com/deisseroth-lab/dask/commit/39d56b91a016edbcbf01b281da2ba433915ee23a

There has been discussion on this topic here, too
https://github.com/dask/dask/issues/5544

I think getting this optimal in the general case is tricky, but perhaps someone has a slick algorithm.

**Comment by phofl** [comment]
the explosion of chunks was fixed in #11273
